src/memray/reporters/assets/temporal_flamegraph.js

```diff
@@ -0,0 +1,299 @@
+        node_objects[nodeIndex].value += bytes;
+        nodeIndex = parent_index_by_child_index[nodeIndex];
+      }
+    }
+  });
+
+  console.log(
+    "total leaked allocations in range: " + node_objects[0].n_allocations
+  );
+  console.log("total leaked bytes in range: " + node_objects[0].value);
+
+  node_objects.forEach((node) => {
+    node.children = node.children.filter((node) => node.n_allocations > 0);
+  });
+
+  return node_objects[0];
+}
+
+function initMemoryGraph(memory_records) {
+  console.log("init memory graph");
+  const time = memory_records.map((a) => new Date(a[0]));
+  const resident_size = memory_records.map((a) => a[1]);
+  const heap_size = memory_records.map((a) => a[2]);
+
+  var resident_size_plot = {
+    x: time,
+    y: resident_size,
+    mode: "lines",
+    name: "Resident size",
+  };
+
+  var heap_size_plot = {
+    x: time,
+    y: heap_size,
+    mode: "lines",
+    name: "Heap size",
+  };
+
+  var plot_data = [resident_size_plot, heap_size_plot];
+  var config = {
+    responsive: true,
+    displayModeBar: false,
+  };
+  var layout = {
+    xaxis: {
+      title: {
+        text: "Time",
+      },
+      rangeslider: {
+        visible: true,
+      },
+    },
+    yaxis: {
+      title: {
+        text: "Memory Size",
+      },
+      tickformat: ".4~s",
+      exponentformat: "B",
+      ticksuffix: "B",
+    },
+  };
+
+  Plotly.newPlot("plot", plot_data, layout, config).then((plot) => {
+    console.assert(active_plot === null);
+    active_plot = plot;
+  });
+}
+
+function showLoadingAnimation() {
+  console.log("showLoadingAnimation");
+  document.getElementById("loading").style.display = "block";
+  document.getElementById("overlay").style.display = "block";
+}
+
+function hideLoadingAnimation() {
+  console.log("hideLoadingAnimation");
+  document.getElementById("loading").style.display = "none";
+  document.getElementById("overlay").style.display = "none";
+}
+
+function refreshFlamegraph(event) {
+  console.log("refreshing flame graph!");
+
+  let request_data = getRangeData(event);
+  console.log("range data: " + request_data);
+
+  if (
+    current_dimensions != null &&
+    JSON.stringify(request_data) === JSON.stringify(current_dimensions)
+  ) {
+    return;
+  }
+
+  console.log("showing loading animation");
+  showLoadingAnimation();
+
+  current_dimensions = request_data;
+
+  console.log("finding range of relevant snapshot");
+
+  let idx0 = 0;
+  let idx1 = memory_records.length - 1;
+
+  if (request_data) {
+    let t0 = new Date(request_data.string1).getTime();
+    for (let i = 0; i < memory_records.length; i++) {
+      if (memory_records[i][0] >= t0) {
+        idx0 = i;
+        break;
+      }
+    }
+
+    idx1 = 0;
+    let t1 = new Date(request_data.string2).getTime();
+    for (let i = memory_records.length - 1; i >= 1; i--) {
+      if (memory_records[i - 1][0] < t1) {
+        idx1 = i;
+        break;
+      }
+    }
+  }
```

Comment:
What about

```
let idx0 = 0;
let idx1 = memory_records.length - 1;

if (request_data) {
  const t0 = new Date(request_data.string1).getTime();
  idx0 = memory_records.findIndex(([t, _]) => t >= t0);

  const t1 = new Date(request_data.string2).getTime();
  idx1 = memory_records.length - 1 - memory_records.slice().reverse().findIndex(([t, _]) => t < t1);
}

idx1 = Math.max(idx1, idx0);
```

?

Reply:
Hm, not opposed to using `findIndex`, but I think that should be something more like:
```js
let idx0 = 0;
let idx1 = memory_records.length - 1;

if (request_data) {
  const t0 = new Date(request_data.string1).getTime();
  const t0_idx = memory_records.findLastIndex((rec) => rec[0] >= t0);
  if (t0_idx != -1) idx0 = t0_idx;

  const t1 = new Date(request_data.string2).getTime();
  const t1_idx = memory_records.findIndex((rec) => rec[0] >= t1);
  if (t1_idx != -1) idx1 = t1_idx;
}
```